[AIRFLOW-2315] Add s3_upload_args to put methods on S3Hook

jbacon opened the pull request:
Hello, please accept my PR to improve the S3Hook.

### JIRA
- [X] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW-2315) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2315

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:
This PR improves the S3Hook by adding a new user-defined JSON extra `s3_upload_args` to allow passing S3 transfer upload arguments (boto3.s3.transfer.S3Transfer.ALLOWED_UPLOAD_ARGS) via embedded JSON. The PR also refactors S3Hook code to reduce duplication. While there are many ways to use this new feature, my specific needs require running Airflow Workers (Celery) across multiple AWS accounts. In this case, to use S3RemoteLogging successfully (from all the workers) I need to pass ExtraArgs to the logger's S3Hook to grant the bucket owner (centralized log bucket) full permission to the log files.

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason: 
New tests are added in `tests/hooks/test_s3_hooks.py` including a `test_load_byes_with_extras` and `test_load_string_with_extras`. The mock_s3 module successfully mocks the ExtraArgs for the virtual S3 resources. 


### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"


### Documentation
- [X] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.


### Code Quality
- [X] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

jgao54 reviewed (commented):
change lgtm! could you add some test coverage?

feng-tao commented on airflow/hooks/S3_hook.py:
```diff
@@ -275,7 +275,8 @@ def load_file(self,
                   replace=False,
-                  encrypt=False):
+                  encrypt=False,
+                  upload_args={}):
```
nit: we shouldn't use {} as default value(https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html) . Should just do ```upload_args=None```

Fokko replied:
This can be quite dangerous indeed if you're not familiar with its behaviour: 
```
MacBook-Pro-van-Fokko:~ fokkodriesprong$ python3
Python 3.7.0 (default, Jun 29 2018, 20:13:13) 
[Clang 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> def some_function(a={}):
...     return a
... 
>>> some_function()
{}
>>> some_function()['lid'] = True
>>> some_function()
{'lid': True}
```

feng-tao commented on airflow/hooks/S3_hook.py:
```diff
@@ -312,7 +312,8 @@ def load_string(self,
                     encrypt=False,
-                    encoding='utf-8'):
+                    encoding='utf-8',
+                    upload_args={}):
```
nit: same

feng-tao reviewed (commented):
CI tests are failing. And I don't have the setup, hence I can't test the change myself as well. But the change lgtm by just looking the code.

Fokko commented on airflow/contrib/operators/ecs_operator.py:
```diff
@@ -52,8 +52,13 @@ class ECSOperator(BaseOperator):
-                 aws_conn_id=None, region_name=None, launch_type='EC2', **kwargs):
+    def __init__(self,
+                 task_definition,
+                 cluster, overrides,
```
Maybe place `overrides` on a separate line as well?

Fokko reviewed (commented):
Care for a review @poulainv?

Commit:
[AIRFLOW-2315] Improve S3Hook with JSON Extras

Adds a new JSON extra key to the S3Hook (`s3_upload_args`) which lets
users pass in S3 upload ExtraArgs in their S3 Connections.
The S3Hook methods `load_string`, `load_file`, and `load_bytes`,
also include a new parameter for overriding the Connections json extra.
Allowed s3 upload arguments can be found by searching:
"boto3.s3.transfer.S3Transfer.ALLOWED_UPLOAD_ARGS"
This commit also refactors S3Hook to reduce code
duplication via a new `_prepare_load` method, which holds
shared logic.
Test converage has been included as well.

Date:      Tue Aug 14 11:55:43 2018 -0700
Changes to be committed:
	modified:   airflow/hooks/S3_hook.py
	modified:   tests/hooks/test_s3_hook.py
